File: src/bmx160_ros/bmx160_ros/bmx160_node.py
import rclpy
import numpy as np
from rclpy.node import Node
from rcl_interfaces.msg import ParameterEvent
from scipy.spatial.transform import Rotation as R
from sensor_msgs.msg import Imu, MagneticField
from bmx160_ros.bmx160_lib import BMX160, GyroRange, AccelRange

class BMX160Node(Node):
    def __init__(self):
        super().__init__('bmx160_node')

        # Parameters
        self.declare_parameter('bus', 1)
        self.declare_parameter('rate_hz', 100.0)
        self.declare_parameter('frame_id', 'imu_link')
        self.declare_parameter('bias_removal', True)
        self.declare_parameter('bias_duration_sec', 2.0)
        self.declare_parameter('accel_bias', [0.0, 0.0, 0.0])
        self.declare_parameter('gyro_bias', [0.0, 0.0, 0.0])
        
        # Sensor Variance Parameters
        self.declare_parameter('accel_variance', [0.001, 0.001, 0.001])
        self.declare_parameter('gyro_variance', [0.0001, 0.0001, 0.0001])
        self.declare_parameter('mag_variance', [1e-3, 1e-3, 1e-3])
        self.declare_parameter('mag_declination_angle', 0.0)
        
        # Magnetometer Calibration Parameters
        self.declare_parameter('mag_bias', [0.0, 0.0, 0.0])
        # Default identity matrix for 3x3 transform, flattened
        self.declare_parameter('mag_transform', [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0])
        
        # IMU Rotation (Mounting)
        self.declare_parameter('imu_rotation_transformation', [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0])

        # IMU Rotation
        self._imu_rot = np.array(self.get_parameter('imu_rotation_transformation').value).reshape(3, 3)

        self.bus = self.get_parameter('bus').value
        self.rate_hz = self.get_parameter('rate_hz').value
        self.frame_id = self.get_parameter('frame_id').value
        self.bias_removal = self.get_parameter('bias_removal').value
        self.bias_duration_sec = self.get_parameter('bias_duration_sec').value

        self.mag_dec_angle = self.get_parameter('mag_declination_angle').value
        self._mag_dec_rot = R.from_euler('xyz', [0, 0, self.mag_dec_angle], degrees=True).as_matrix()

        # Initialize Sensor
        self.imu = BMX160(self.bus)
        if self.imu.begin():
            self.get_logger().info('✅ BMX160 init OK')
    
            # Configure gyro range (±500°/s is good for most applications)
            self.imu.set_gyro_range(GyroRange.DPS_500)
            
            # Configure accel range (±4g is good for most applications)
            self.imu.set_accel_range(AccelRange.G_4)
            
            self.get_logger().info(f'✅ Gyro configured: {GyroRange.DPS_500} sensitivity')
            self.get_logger().info(f'✅ Accel configured: {AccelRange.G_4} sensitivity')

        else:
            self.get_logger().error('❌ BMX160 init FAILED (check I2C)')
            raise RuntimeError('BMX160 init failed')

        # Publishers
        self.pub_imu = self.create_publisher(Imu, 'imu/data', 10)
        self.pub_mag = self.create_publisher(MagneticField, 'imu/mag', 10)

        # Calibration State
        self.calibrating = self.bias_removal
        self.calibration_samples = {
            'accel': [],
            'gyro': []
        }
        self.calibration_end_time = None
        
        # Biases
        self._accel_bias = np.array(self.get_parameter('accel_bias').value)
        self._gyro_bias = np.array(self.get_parameter('gyro_bias').value)

        # Variances
        self._accel_cov = np.diag(self.get_parameter('accel_variance').value).flatten()
        self._gyro_cov = np.diag(self.get_parameter('gyro_variance').value).flatten()
        self._mag_cov = np.diag(self.get_parameter('mag_variance').value).flatten()
        
        # Magnetometer Calibration (Loaded from params initially)
        self._mag_bias = np.array(self.get_parameter('mag_bias').value)
        self._mag_transform = np.array(self.get_parameter('mag_transform').value).reshape(3, 3)
        

        # Parameter Callback
        self.sub_params = self.create_subscription(
            ParameterEvent,
            '/parameter_events',
            self.on_param_event,
            10
        )

        # Timer
        period = 1.0 / self.rate_hz if self.rate_hz > 0 else 0.01
        self.timer = self.create_timer(period, self.tick)
        
        if self.calibrating:
             self.get_logger().info(f'⚠ Starting Bias Calibration ({self.bias_duration_sec}s). Keep vehicle LEVEL and STATIONARY.')
             self.calibration_end_time = self.get_clock().now().nanoseconds / 1e9 + self.bias_duration_sec

    def tick(self):
        try:
            d = self.imu.get_all_data()
            if not d:
                return

            # Raw Data
            # Magnetometer: uT
            mx, my, mz = float(d[0]), float(d[1]), float(d[2])
            
            # Mapping
            ax_ms2, ay_ms2, az_ms2 = float(d[6]), float(d[7]), float(d[8])
            gx_dps, gy_dps, gz_dps = float(d[3]), float(d[4]), float(d[5])
            
            gx_rad = np.deg2rad(gx_dps)
            gy_rad = np.deg2rad(gy_dps)
            gz_rad = np.deg2rad(gz_dps)

            if self.calibrating:
                now_sec = self.get_clock().now().nanoseconds / 1e9
                if now_sec < self.calibration_end_time:
                    self.calibration_samples['accel'].append([ax_ms2, ay_ms2, az_ms2])
                    self.calibration_samples['gyro'].append([gx_rad, gy_rad, gz_rad])
                else:
                    # Finish calibration
                    accel_data = np.array(self.calibration_samples['accel'])
                    gyro_data = np.array(self.calibration_samples['gyro'])
                    
                    if len(accel_data) > 0:
                        self._gyro_bias = np.mean(gyro_data, axis=0)

                        # Accel bias: assume Z is Gravity (9.80665)
                        # We want readings to be [0, 0, +g] when level.
                        # measured = true + bias => bias = measured - true
                        
                        gravity = self._imu_rot @ np.array([0.0, 0.0, 9.80665])
                        a = accel_data + gravity
                        self._accel_bias = np.mean(a, axis=0)
                        
                        self.get_logger().info(f'Calibration Done. Gyro Bias: {self._gyro_bias}, Accel Bias: {self._accel_bias}')
                    else:
                        self.get_logger().warn('Calibration failed: no data collected.')
                    
                    self.calibrating = False
                return

            # Apply Logic
            # IMU Message
            imu_msg = Imu()
            imu_msg.header.stamp = self.get_clock().now().to_msg()
            imu_msg.header.frame_id = self.frame_id

            # Apply bias removal (Sensor Frame)
            ax_calib = ax_ms2 - self._accel_bias[0]
            ay_calib = ay_ms2 - self._accel_bias[1]
            az_calib = az_ms2 - self._accel_bias[2]
            
            gx_calib = gx_rad - self._gyro_bias[0]
            gy_calib = gy_rad - self._gyro_bias[1]
            gz_calib = gz_rad - self._gyro_bias[2]
            
            # Apply Rotation (Mounting -> Body)
            accel_vec = np.array([ax_calib, ay_calib, az_calib])
            gyro_vec = np.array([gx_calib, gy_calib, gz_calib])
            
            accel_body = self._imu_rot @ accel_vec
            gyro_body = self._imu_rot @ gyro_vec

            imu_msg.linear_acceleration.x = accel_body[0]
            imu_msg.linear_acceleration.y = accel_body[1]
            imu_msg.linear_acceleration.z = accel_body[2]
            
            imu_msg.angular_velocity.x = gyro_body[0]
            imu_msg.angular_velocity.y = gyro_body[1]
            imu_msg.angular_velocity.z = gyro_body[2]
            
            # Set covariance
            imu_msg.orientation_covariance[0] = -1.0 # Orientation not provided
            imu_msg.linear_acceleration_covariance = self._accel_cov
            imu_msg.angular_velocity_covariance = self._gyro_cov
            
            self.pub_imu.publish(imu_msg)

            # Magnetic Field Message
            mag_msg = MagneticField()
            mag_msg.header.stamp = imu_msg.header.stamp # Sync
            mag_msg.header.frame_id = self.frame_id
            mag_msg.magnetic_field_covariance = self._mag_cov
            
            # Calibration Logic from user request
            B = np.array([mx, my, mz])
            
            # Apply Rotation (Mounting -> Body)
            B_body = self._imu_rot @ B

            # Soft Iron / Bias (Sensor Frame)
            B_bias = self._imu_rot @ self._mag_bias
            B_calib = (B_body - B_bias) @ self._mag_transform.T

            B_true_north = self._mag_dec_rot.T @ B_calib

            mag_msg.magnetic_field.x = B_true_north[0]
            mag_msg.magnetic_field.y = B_true_north[1]
            mag_msg.magnetic_field.z = B_true_north[2]

            self.pub_mag.publish(mag_msg)

        except Exception as e:
            self.get_logger().warn(f'read/publish error: {e}')

    def on_param_event(self, event):
        for changed in event.changed_parameters:
            if changed.name == "mag_bias":
                if len(changed.value.double_array_value) == 3:
                     self._mag_bias = np.array(changed.value.double_array_value)
                     self.get_logger().info(f"Updated mag_bias: {self._mag_bias}")
            elif changed.name == "mag_transform":
                if len(changed.value.double_array_value) == 9:
                    self._mag_transform = np.array(changed.value.double_array_value).reshape(3, 3)
                    self.get_logger().info(f"Updated mag_transform parameters")
            elif changed.name == "imu_rotation_transformation":
                 if len(changed.value.double_array_value) == 9:
                    self._imu_rot = np.array(changed.value.double_array_value).reshape(3, 3)
                    self.get_logger().info(f"Updated imu_rotation_transformation")
            elif changed.name == "accel_bias":
                if len(changed.value.double_array_value) == 3:
                     self._accel_bias = np.array(changed.value.double_array_value)
                     self.get_logger().info(f"Updated accel_bias: {self._accel_bias}")
            elif changed.name == "gyro_bias":
                if len(changed.value.double_array_value) == 3:
                     self._gyro_bias = np.array(changed.value.double_array_value)
                     self.get_logger().info(f"Updated gyro_bias: {self._gyro_bias}")
            elif changed.name == "accel_variance":
                if len(changed.value.double_array_value) == 3:
                     self._accel_cov = np.diag(changed.value.double_array_value).flatten()
                     self.get_logger().info("Updated accel_variance")
            elif changed.name == "gyro_variance":
                if len(changed.value.double_array_value) == 3:
                     self._gyro_cov = np.diag(changed.value.double_array_value).flatten()
                     self.get_logger().info("Updated gyro_variance")
            elif changed.name == "mag_variance":
                if len(changed.value.double_array_value) == 3:
                     self._mag_cov = np.diag(changed.value.double_array_value).flatten()
                     self.get_logger().info("Updated mag_variance")
            elif changed.name == "mag_declination_angle":
                self._mag_cov = changed.value.double_value
                self.get_logger().info("Updated mag_declination_angle")

    def destroy_node(self):
        super().destroy_node()

def main(args=None):
    rclpy.init(args=args)
    node = BMX160Node()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/bmx160_ros/src/bmx160_node.cpp
#include <chrono>
#include <memory>
#include <vector>
#include <string>
#include <cmath>
#include <numeric>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/imu.hpp"
#include "sensor_msgs/msg/magnetic_field.hpp"
#include "rcl_interfaces/msg/parameter_event.hpp"

#include "bmx160_ros/bmx160.hpp"

using namespace std::chrono_literals;

namespace bmx160_ros {

class BMX160Node : public rclcpp::Node {
public:
    BMX160Node() : Node("bmx160_node") {
        // Parameters
        this->declare_parameter("bus", 1);
        this->declare_parameter("rate_hz", 100.0);
        this->declare_parameter("frame_id", "imu_link");
        this->declare_parameter("bias_removal", true);
        this->declare_parameter("bias_duration_sec", 2.0);
        this->declare_parameter("accel_bias", std::vector<double>{0.0, 0.0, 0.0});
        this->declare_parameter("gyro_bias", std::vector<double>{0.0, 0.0, 0.0});
        
        this->declare_parameter("accel_variance", std::vector<double>{0.001, 0.001, 0.001});
        this->declare_parameter("gyro_variance", std::vector<double>{0.0001, 0.0001, 0.0001});
        this->declare_parameter("mag_variance", std::vector<double>{1e-3, 1e-3, 1e-3});
        this->declare_parameter("mag_declination_angle", 0.0);
        
        this->declare_parameter("mag_bias", std::vector<double>{0.0, 0.0, 0.0});
        this->declare_parameter("mag_transform", std::vector<double>{1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0});
        this->declare_parameter("imu_rotation_transformation", std::vector<double>{1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0});

        loadParameters();

        // Initialize Sensor
        imu_ = std::make_unique<BMX160>(bus_);
        if (imu_->begin()) {
            RCLCPP_INFO(this->get_logger(), "✅ BMX160 init OK");
            imu_->setGyroRange(GyroRange::DPS_500);
            imu_->setAccelRange(AccelRange::G_4);
        } else {
            RCLCPP_ERROR(this->get_logger(), "❌ BMX160 init FAILED (check I2C)");
            throw std::runtime_error("BMX160 init failed");
        }

        // Publishers
        pub_imu_ = this->create_publisher<sensor_msgs::msg::Imu>("imu/data", 10);
        pub_mag_ = this->create_publisher<sensor_msgs::msg::MagneticField>("imu/mag", 10);

        // Calibration State
        calibrating_ = bias_removal_;
        if (calibrating_) {
            RCLCPP_INFO(this->get_logger(), "⚠ Starting Bias Calibration (%.1fs). Keep vehicle LEVEL and STATIONARY.", bias_duration_sec_);
            calibration_end_time_ = this->now().seconds() + bias_duration_sec_;
        }

        // Parameter Callback
        param_event_sub_ = this->create_subscription<rcl_interfaces::msg::ParameterEvent>(
            "/parameter_events", 10, std::bind(&BMX160Node::onParamEvent, this, std::placeholders::_1));

        // Timer
        double period = 1.0 / rate_hz_;
        timer_ = this->create_wall_timer(std::chrono::duration<double>(period), std::bind(&BMX160Node::tick, this));
    }

private:
    void loadParameters() {
        bus_ = this->get_parameter("bus").as_int();
        rate_hz_ = this->get_parameter("rate_hz").as_double();
        frame_id_ = this->get_parameter("frame_id").as_string();
        bias_removal_ = this->get_parameter("bias_removal").as_bool();
        bias_duration_sec_ = this->get_parameter("bias_duration_sec").as_double();
        mag_dec_angle_ = this->get_parameter("mag_declination_angle").as_double();

        auto ab = this->get_parameter("accel_bias").as_double_array();
        for (int i = 0; i < 3; ++i) accel_bias_[i] = ab[i];

        auto gb = this->get_parameter("gyro_bias").as_double_array();
        for (int i = 0; i < 3; ++i) gyro_bias_[i] = gb[i];

        auto av = this->get_parameter("accel_variance").as_double_array();
        for (int i = 0; i < 3; ++i) accel_variance_[i] = av[i];

        auto gv = this->get_parameter("gyro_variance").as_double_array();
        for (int i = 0; i < 3; ++i) gyro_variance_[i] = gv[i];

        auto mv = this->get_parameter("mag_variance").as_double_array();
        for (int i = 0; i < 3; ++i) mag_variance_[i] = mv[i];

        auto mb = this->get_parameter("mag_bias").as_double_array();
        for (int i = 0; i < 3; ++i) mag_bias_[i] = mb[i];

        auto mt = this->get_parameter("mag_transform").as_double_array();
        for (int i = 0; i < 9; ++i) mag_transform_[i/3][i%3] = mt[i];

        auto ir = this->get_parameter("imu_rotation_transformation").as_double_array();
        for (int i = 0; i < 9; ++i) imu_rot_[i/3][i%3] = ir[i];

        updateMagDeclinationMatrix();
    }

    void updateMagDeclinationMatrix() {
        double rad = mag_dec_angle_ * M_PI / 180.0;
        mag_dec_rot_[0][0] = std::cos(rad);
        mag_dec_rot_[0][1] = -std::sin(rad);
        mag_dec_rot_[0][2] = 0;
        mag_dec_rot_[1][0] = std::sin(rad);
        mag_dec_rot_[1][1] = std::cos(rad);
        mag_dec_rot_[1][2] = 0;
        mag_dec_rot_[2][0] = 0;
        mag_dec_rot_[2][1] = 0;
        mag_dec_rot_[2][2] = 1;
    }

    void tick() {
        auto data = imu_->getAllData();
        if (!data) return;

        if (calibrating_) {
            double now = this->now().seconds();
            if (now < calibration_end_time_) {
                calib_samples_accel_.push_back({data->accel[0], data->accel[1], data->accel[2]});
                calib_samples_gyro_.push_back({data->gyro[0] * (float)M_PI / 180.0f, 
                                               data->gyro[1] * (float)M_PI / 180.0f, 
                                               data->gyro[2] * (float)M_PI / 180.0f});
            } else {
                finishCalibration();
            }
            return;
        }

        auto imu_msg = sensor_msgs::msg::Imu();
        imu_msg.header.stamp = this->now();
        imu_msg.header.frame_id = frame_id_;

        // Apply bias (Sensor Frame)
        double ax = data->accel[0] - accel_bias_[0];
        double ay = data->accel[1] - accel_bias_[1];
        double az = data->accel[2] - accel_bias_[2];

        double gx = (data->gyro[0] * M_PI / 180.0) - gyro_bias_[0];
        double gy = (data->gyro[1] * M_PI / 180.0) - gyro_bias_[1];
        double gz = (data->gyro[2] * M_PI / 180.0) - gyro_bias_[2];

        // Apply Rotation (Mounting -> Body)
        double accel_body[3] = {0, 0, 0};
        double gyro_body[3] = {0, 0, 0};
        for (int i = 0; i < 3; ++i) {
            accel_body[i] = imu_rot_[i][0] * ax + imu_rot_[i][1] * ay + imu_rot_[i][2] * az;
            gyro_body[i] = imu_rot_[i][0] * gx + imu_rot_[i][1] * gy + imu_rot_[i][2] * gz;
        }

        imu_msg.linear_acceleration.x = accel_body[0];
        imu_msg.linear_acceleration.y = accel_body[1];
        imu_msg.linear_acceleration.z = accel_body[2];
        imu_msg.angular_velocity.x = gyro_body[0];
        imu_msg.angular_velocity.y = gyro_body[1];
        imu_msg.angular_velocity.z = gyro_body[2];

        imu_msg.orientation_covariance[0] = -1.0;
        for (int i = 0; i < 3; ++i) {
            imu_msg.linear_acceleration_covariance[i*4] = accel_variance_[i];
            imu_msg.angular_velocity_covariance[i*4] = gyro_variance_[i];
        }

        pub_imu_->publish(imu_msg);

        // Magnetometer
        auto mag_msg = sensor_msgs::msg::MagneticField();
        mag_msg.header.stamp = imu_msg.header.stamp;
        mag_msg.header.frame_id = frame_id_;

        double mx = data->mag[0];
        double my = data->mag[1];
        double mz = data->mag[2];

        // Apply Rotation (Mounting -> Body)
        double mx_body = imu_rot_[0][0] * mx + imu_rot_[0][1] * my + imu_rot_[0][2] * mz;
        double my_body = imu_rot_[1][0] * mx + imu_rot_[1][1] * my + imu_rot_[1][2] * mz;
        double mz_body = imu_rot_[2][0] * mx + imu_rot_[2][1] * my + imu_rot_[2][2] * mz;

        // Apply Mag Bias (In Body frame matching Python logic)
        double mb_body[3] = {0, 0, 0};
        for (int i = 0; i < 3; ++i) {
            mb_body[i] = imu_rot_[i][0] * mag_bias_[0] + imu_rot_[i][1] * mag_bias_[1] + imu_rot_[i][2] * mag_bias_[2];
        }

        double dx = mx_body - mb_body[0];
        double dy = my_body - mb_body[1];
        double dz = mz_body - mb_body[2];

        // Apply Soft Iron (mag_transform)
        double m_calib[3] = {0, 0, 0};
        for (int i = 0; i < 3; ++i) {
            m_calib[i] = mag_transform_[i][0] * dx + mag_transform_[i][1] * dy + mag_transform_[i][2] * dz;
        }

        // Apply Declination (Transpose of mag_dec_rot per Python code)
        double m_true[3] = {0, 0, 0};
        for (int i = 0; i < 3; ++i) {
            m_true[i] = mag_dec_rot_[0][i] * m_calib[0] + mag_dec_rot_[1][i] * m_calib[1] + mag_dec_rot_[2][i] * m_calib[2];
        }

        mag_msg.magnetic_field.x = m_true[0];
        mag_msg.magnetic_field.y = m_true[1];
        mag_msg.magnetic_field.z = m_true[2];
        for (int i = 0; i < 3; ++i) mag_msg.magnetic_field_covariance[i*4] = mag_variance_[i];

        pub_mag_->publish(mag_msg);
    }

    void finishCalibration() {
        if (calib_samples_accel_.empty()) {
            RCLCPP_WARN(this->get_logger(), "Calibration failed: no data collected.");
            calibrating_ = false;
            return;
        }

        // Gyro bias is simple mean
        std::vector<double> sum_gyro = {0, 0, 0};
        for (const auto& s : calib_samples_gyro_) {
            sum_gyro[0] += s[0]; sum_gyro[1] += s[1]; sum_gyro[2] += s[2];
        }
        for (int i = 0; i < 3; ++i) gyro_bias_[i] = sum_gyro[i] / calib_samples_gyro_.size();

        // Accel bias assumes Z is gravity
        // gravity_sensor = imu_rot^T * [0, 0, 9.80665]
        double gravity_sensor[3] = {0, 0, 0};
        for (int i = 0; i < 3; ++i) gravity_sensor[i] = imu_rot_[0][i] * 0 + imu_rot_[1][i] * 0 + imu_rot_[2][i] * 9.80665;

        // bias = measured - gravity
        std::vector<double> sum_accel = {0, 0, 0};
        for (const auto& s : calib_samples_accel_) {
            sum_accel[0] += s[0] + gravity_sensor[0];
            sum_accel[1] += s[1] + gravity_sensor[1];
            sum_accel[2] += s[2] + gravity_sensor[2];
        }
        for (int i = 0; i < 3; ++i) accel_bias_[i] = sum_accel[i] / calib_samples_accel_.size();

        RCLCPP_INFO(this->get_logger(), "Calibration Done. Gyro Bias: [%.4f, %.4f, %.4f], Accel Bias: [%.4f, %.4f, %.4f]",
                    gyro_bias_[0], gyro_bias_[1], gyro_bias_[2], accel_bias_[0], accel_bias_[1], accel_bias_[2]);

        calibrating_ = false;
        calib_samples_accel_.clear();
        calib_samples_gyro_.clear();
    }

    void onParamEvent(const rcl_interfaces::msg::ParameterEvent::SharedPtr event) {
        if (event->node != this->get_fully_qualified_name()) return;

        for (auto& param : event->changed_parameters) {
            if (param.name == "mag_bias") {
                auto val = param.value.double_array_value;
                if (val.size() == 3) for (int i = 0; i < 3; ++i) mag_bias_[i] = val[i];
            } else if (param.name == "mag_transform") {
                auto val = param.value.double_array_value;
                if (val.size() == 9) for (int i = 0; i < 9; ++i) mag_transform_[i/3][i%3] = val[i];
            } else if (param.name == "imu_rotation_transformation") {
                auto val = param.value.double_array_value;
                if (val.size() == 9) for (int i = 0; i < 9; ++i) imu_rot_[i/3][i%3] = val[i];
            } else if (param.name == "accel_bias") {
                auto val = param.value.double_array_value;
                if (val.size() == 3) for (int i = 0; i < 3; ++i) accel_bias_[i] = val[i];
            } else if (param.name == "gyro_bias") {
                auto val = param.value.double_array_value;
                if (val.size() == 3) for (int i = 0; i < 3; ++i) gyro_bias_[i] = val[i];
            } else if (param.name == "mag_declination_angle") {
                mag_dec_angle_ = param.value.double_value;
                updateMagDeclinationMatrix();
            }
        }
    }

    std::unique_ptr<BMX160> imu_;
    rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr pub_imu_;
    rclcpp::Publisher<sensor_msgs::msg::MagneticField>::SharedPtr pub_mag_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Subscription<rcl_interfaces::msg::ParameterEvent>::SharedPtr param_event_sub_;

    int bus_;
    double rate_hz_;
    std::string frame_id_;
    bool bias_removal_;
    double bias_duration_sec_;
    double mag_dec_angle_;

    double accel_bias_[3] = {0, 0, 0};
    double gyro_bias_[3] = {0, 0, 0};
    double mag_bias_[3] = {0, 0, 0};
    double accel_variance_[3], gyro_variance_[3], mag_variance_[3];
    double mag_transform_[3][3], imu_rot_[3][3], mag_dec_rot_[3][3];

    bool calibrating_;
    double calibration_end_time_;
    std::vector<std::vector<double>> calib_samples_accel_, calib_samples_gyro_;
};

} // namespace bmx160_ros

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<bmx160_ros::BMX160Node>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
